_posts/2016-04-19-MIR.md

```diff
@@ -0,0 +1,725 @@
+
+```rust
+let mut iterator = vec.into_iter();
+loop {
+    match iterator.next() {
+        Some(elem) => process(elem),
+        None => break,
+    }
+}
+```
+
+We've already eliminated two constructs (`for` loops and `while`
+loops), but we can go further still. Let's turn from loops for a bit
+to look at the method calls that we see. In Rust, method calls like
+`vec.into_iter()` and `iterator.next()` are also a kind of
+[syntactic sugar][ss]. These particular methods are defined in traits,
+which are basically pre-defined interfaces.  For example, `into_iter`
+is a method in the `IntoIterator` trait.  Types which can be converted
+into iterators implement that trait and define how the `into_iter`
+method works for them. Similarly, `next` is defined in the `Iterator`
+trait. When you write a method call like `iterator.next()`, the Rust
+compiler automatically figures out which trait the method belongs to
+based on the type of the `iterator` and the set of traits in
+scope. But if we prefer to be more explicit, we could instead invoke
+the methods in the trait directly, using function call syntax:
+
+```rust
+// Rather than `vec.into_iter()`, we are calling
+// the function `IntoIterator::into_iter`. This is
+// exactly equivalent, just more explicit.
+let mut iterator = IntoIterator::into_iter(vec);
+loop {
+    // Similarly, `iterator.next()` can be rewritten
+    // to make clear which trait the `next` method
+    // comes from. We see here that the `.` notation
+    // was also adding an implicit mutable reference,
+    // which is now made explicit.
+    match Iterator::next(&mut iterator) {
+        Some(elem) => process(elem),
+        None => break,
+    }
+}
+```
+
+At this point, we've managed to reduce the set of language features
+for our little fragment quite a bit: we now only use `loop` loops and
+we don't use method calls. But MIR in fact simplifies things even
+more: it removes the very notion of `loop` and `break` keywords by
+using a **control-flow graph**.
+
+### Control-flow graphs
+
+In the examples so far, I was presenting MIR as a gradual
+simplification of Rust programs into other Rust programs. In reality,
+though, MIR is based on the notion of a
+[control-flow graph (CFG)][CFG]. If you've ever used a flow-chart,
+then the concept of a control-flow graph will be pretty familiar to
+you. It's a representation of your program that exploses the
+underlying control flow in a very clear way.
+
+When we build a control-flow graph, explicit control-flow constructs
+like the `loop` or `break` keywords disappear. Instead, they are
+expressed by paths in the graph. So a loop simply appears as a cycle
+in the graph, and the `break` keyword translates into a path out of
+that cycle. To put it another way, through the CFG, the MIR
+representation effectively has a general `goto`, and so we can express
+the `loop` using that.
+
+Here is the running example from the preivous section, expressed as a
+control-flow graph:
```

Comment:
I would like to see this as pseudo code with `goto`s and labels (just this one, not all the CFGs, and as well as the diagram). Maybe I'm just weird, but for me it is easier to make the transition from code to graph with that extra step.